File: exercício 3/js/features/step_definitions/cat_breeds.js
const axios = require('axios');
const assert = require('assert');
const { Given, When, Then } = require('@cucumber/cucumber');

const BASE_URL = 'https://catfact.ninja';
let response;

Given('I am making a GET request to the valid {string} endpoint', function (endpoint) {
  this.endpoint = `${BASE_URL}/breeds`;
  response = null;
});

When('I send the valid request', async function () {
  response = await axios.get(this.endpoint);
});

Then('I should receive a succesful {string} response', function (statusCode) {
  assert.strictEqual(response.status, 200, `Status code should be ${statusCode}, not ${response.status}`);
});

Then('the response body should contain a list of cat breeds', function () {
  const data = response.data;
  assert.ok(data.hasOwnProperty('data'), 'Response body does not contain valid "data"');
  assert.ok(Array.isArray(data.data), 'Response body invalid');
  assert.ok(data.data.length > 0, 'Response body is empty');
});

Given('I am making a GET request to the invalid {string} endpoint', function (endpoint) {
  this.endpoint = `${BASE_URL}/invalid-route`;
  response = null;
});

When('I send the invalid request', async function () {
  try {
    response = await axios.get(this.endpoint);
  } catch (error) {
    response = error.response;
  }
});

Then('I should receive a unsuccesful {string} response', function (statusCode) {
  assert.strictEqual(response.status, 404, `Status code should be ${statusCode}, not ${response.status}`);
});

File: exercício 3/python/features/steps/cat_breeds.py
import requests
from behave import *

BASE_URL = "https://catfact.ninja"

@given('I am making a GET request to the "/breeds" endpoint')
def step_given_url(context):
    context.url = f"{BASE_URL}/breeds"
    context.response = None

@when('I send the request')
def step_when_send_request(context):
    context.response = requests.get(context.url)

@then('I should receive a "200 OK" response')
def step_then_status_code(context):
    assert context.response.status_code == 200

@then('the response body should contain a list of cat breeds')
def step_then_response_body(context):
    response = context.response.json()
    assert 'data' in response
    assert isinstance(response['data'], list)
    assert len(response['data']) > 0
    
####################################################################

@given('I am making a GET request to the "/invalid-route" endpoint')
def step_given_invalid_url(context):
    context.url = f"{BASE_URL}/invalid-route"
    context.response = None

@when('I send the invalid request')
def step_when_send_invalid_request(context):
    context.response = requests.get(context.url)

@then('I should receive a "404 Not Found" response')
def step_then_invalid_status_code(context):
    assert context.response.status_code == 404